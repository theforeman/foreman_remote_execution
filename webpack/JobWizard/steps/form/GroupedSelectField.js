import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  SelectOption,
  Select,
  SelectGroup,
  SelectVariant,
  FormGroup,
} from '@patternfly/react-core';

export const GroupedSelectField = ({
  label,
  fieldId,
  groups,
  selected,
  setSelected,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const onSelect = selection => {
    setIsOpen(false);
    setSelected(selection);
  };

  const onClear = () => {
    onSelect(null);
  };

  const options = groups.map((group, groupIndex) => (
    <SelectGroup key={groupIndex} label={group.groupLabel}>
      {group.options.map((option, optionIndex) => (
        <SelectOption
          key={optionIndex}
          value={option.label}
          onClick={() => onSelect(option.value)}
        />
      ))}
    </SelectGroup>
  ));

  const onFilter = evt => {
    const textInput = evt?.target?.value || '';
    if (textInput === '') {
      return options;
    }
    return options
      .map(group => {
        const filteredGroup = React.cloneElement(group, {
          children: group.props.children.filter(item =>
            item.props.value.toLowerCase().includes(textInput.toLowerCase())
          ),
        });
        if (filteredGroup.props.children.length > 0) return filteredGroup;
        return null;
      })
      .filter(newGroup => newGroup);
  };

  return (
    <FormGroup label={label} fieldId={fieldId}>
      <Select
        isGrouped
        variant={SelectVariant.typeahead}
        onToggle={setIsOpen}
        onFilter={onFilter}
        isOpen={isOpen}
        onSelect={() => null}
        selections={selected}
        className="without_select2"
        onClear={onClear}
      >
        {options}
      </Select>
    </FormGroup>
  );
};

GroupedSelectField.propTypes = {
  label: PropTypes.string.isRequired,
  fieldId: PropTypes.string.isRequired,
  groups: PropTypes.array,
  selected: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  setSelected: PropTypes.func.isRequired,
};

GroupedSelectField.defaultProps = {
  groups: [],
  selected: null,
};
