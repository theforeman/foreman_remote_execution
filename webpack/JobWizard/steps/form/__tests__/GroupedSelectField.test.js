import React from 'react';
import * as patternfly from '@patternfly/react-core';
import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { GroupedSelectField } from '../GroupedSelectField';

jest.spyOn(patternfly, 'Select');
jest.spyOn(patternfly, 'SelectOption');
patternfly.Select.mockImplementation(props => <div>{props}</div>);
patternfly.SelectOption.mockImplementation(props => <div>{props}</div>);

const fixtures = {
  'renders with props': {
    label: 'grouped select',
    fieldId: 'field-id',
    groups: [
      {
        groupLabel: 'Ansible',
        options: [
          {
            label: 'Ansible Roles - Ansible Default',
            value: 168,
          },
          {
            label: 'Ansible Roles - Install from git',
            value: 170,
          },
        ],
      },
    ],
    selected: 170,
    setSelected: jest.fn(),
  },
};

describe('GroupedSelectField', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(GroupedSelectField, fixtures));
});
