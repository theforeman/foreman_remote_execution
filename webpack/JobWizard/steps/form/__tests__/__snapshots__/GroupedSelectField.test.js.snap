// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupedSelectField rendering renders with props 1`] = `
<FormGroup
  fieldId="field-id"
  label="grouped select"
>
  <mockConstructor
    className="without_select2"
    isGrouped={true}
    isOpen={false}
    onClear={[Function]}
    onFilter={[Function]}
    onSelect={[Function]}
    onToggle={[Function]}
    selections={170}
    variant="typeahead"
  >
    <SelectGroup
      key="0"
      label="Ansible"
    >
      <mockConstructor
        key="0"
        onClick={[Function]}
        value="Ansible Roles - Ansible Default"
      />
      <mockConstructor
        key="1"
        onClick={[Function]}
        value="Ansible Roles - Install from git"
      />
    </SelectGroup>
  </mockConstructor>
</FormGroup>
`;
